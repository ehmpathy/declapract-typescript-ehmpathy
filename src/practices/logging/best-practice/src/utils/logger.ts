import { generateLogMethods, LOG_LEVEL } from 'simple-leveled-log-methods';

/*
  purpose of logger:
  - distinguish different levels of logs
    - specifies importance
    - allows filtering which to send
  - defines standard format of log messages
  - sends to standard transport
*/

/*
  define the minimal log level
*/
const defaultLogLevel = process.env.AWS_LAMBDA_FUNCTION_NAME
  ? LOG_LEVEL.DEBUG // if AWS_LAMBA_FUNCTION_NAME is set, then we're in lambda env and should transport all messages to console
  : LOG_LEVEL.INFO; // otherwise, we're running locally and should only show info and above
const minimalLogLevel = (process.env.LOG_LEVEL as LOG_LEVEL) || defaultLogLevel; // use the log level specified or fallback to default if none specified

/*
  define the log methods
*/
export const log = generateLogMethods({ minimalLogLevel });
