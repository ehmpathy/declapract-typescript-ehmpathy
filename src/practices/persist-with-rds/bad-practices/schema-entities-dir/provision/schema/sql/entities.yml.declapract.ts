import { FileCheckType, FileFixFunction } from 'declapract';

export const check = FileCheckType.EXISTS;

export const fix: FileFixFunction = (_, context) => {
  return {
    // rename file to "domain.yml"
    relativeFilePath: context.relativeFilePath.replace(
      /entities\.yml$/,
      'domain.yml',
    ),
  };
};
