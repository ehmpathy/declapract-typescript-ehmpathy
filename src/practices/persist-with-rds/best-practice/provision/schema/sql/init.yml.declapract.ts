import { FileCheckType } from 'declapract';

export const check = FileCheckType.CONTAINS; // contains since the user may have exclusions in that "for" object
