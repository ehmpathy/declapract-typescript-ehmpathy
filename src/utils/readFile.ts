import fs from 'fs';
import util from 'util';

export const readFile = async (filePath: string) =>
  util.promisify(fs.readFile)(filePath, 'utf-8');
