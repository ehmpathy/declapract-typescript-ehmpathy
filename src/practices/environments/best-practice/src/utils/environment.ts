import { createIsOfEnum } from 'type-fns';

import { log } from './logger';

export enum Stage {
  PRODUCTION = 'prod',
  DEVELOPMENT = 'dev',
  TEST = 'test',
}
export const isOfStage = createIsOfEnum(Stage);

/**
 * ensure that the server is on UTC timezone
 *
 * why?
 * - non UTC timezone usage causes consistency problems and takes a while to track down
 * - by warning and correcting if the server the code runs on in is not in UTC, we avoid these issues
 * - instead, users should push down converting from UTC into the users TZ as far as possible
 */
const TIMEZONE = process.env.TZ;
if (TIMEZONE !== 'UTC') {
  log.warn(
    'env.TZ is not set to UTC. this can cause issues. updating this on your behalf',
    { found: TIMEZONE, desire: 'UTC' },
  );
  process.env.TZ = 'UTC';
}

/**
 * this allows us to infer what the stage should be in environments that do not have STAGE specified
 * - e.g., when running locally
 * - e.g., when running tests
 */
const inferStageFromNodeEnv = () => {
  const nodeEnv = process.env.NODE_ENV; // default to test if not defined
  if (!nodeEnv) throw new Error('process.env.NODE_ENV must be defined');
  if (nodeEnv === 'production') return Stage.PRODUCTION;
  if (nodeEnv === 'development') return Stage.DEVELOPMENT;
  if (nodeEnv === 'test') return Stage.TEST;
  throw new Error(`unexpected nodeEnv '${nodeEnv}'`);
};

/**
 * a method that exposes relevant environmental variables in a standard way
 */
const getEnvironment = () => {
  const stage = process.env.STAGE ?? inferStageFromNodeEnv(); // figure it out from NODE_ENV if not explicitly defined
  if (!stage) throw new Error('process.env.STAGE must be defined');
  if (!isOfStage(stage)) throw new Error(`invalid stage defined '${stage}'`);
  return { stage };
};

// export stage immediately, since it does not change; // todo: replace this with env.access
export const { stage } = getEnvironment();

// export service client stage
export const serviceClientStage =
  stage === Stage.PRODUCTION ? Stage.PRODUCTION : Stage.DEVELOPMENT; // i.e., if its prod, hit prod. otherwise, dev

// export whether we were asked to run locally; // todo, replace this with env.server
export const locally = process.env.LOCALLY === 'true'; // whether we want to acceptance test locally or deployed lambda
