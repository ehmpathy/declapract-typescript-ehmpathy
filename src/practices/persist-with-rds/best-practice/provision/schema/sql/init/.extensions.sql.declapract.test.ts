import { readFile } from '../../../../../../../utils/readFile';
import { fix } from './.extensions.sql.declapract';

describe('.extensions.sql fix', () => {
  it('should be able to fix a file with custom extensions correctly', async () => {
    const declaredFileContents = await readFile(`${__dirname}/.extensions.sql`); // grab the declared contents
    const exampleFoundContents = `
/**
 NOTE: because db owner is not allowed to create extensions, these changes are applied manually

  NOTE: for the integration test db, these are applied automatically upon spinning up the docker instance
*/
CREATE EXTENSION IF NOT EXISTS "uuid-ossp";
CREATE EXTENSION IF NOT EXISTS "pgcrypto";
CREATE EXTENSION IF NOT EXISTS "cube";
CREATE EXTENSION IF NOT EXISTS "earthdistance";
    `;
    const { contents: fixedContents } = await fix(exampleFoundContents, {
      declaredFileContents,
    } as any);
    expect(fixedContents).toContain('CREATE EXTENSION IF NOT EXISTS "cube";');
    expect(fixedContents).toContain(
      'CREATE EXTENSION IF NOT EXISTS "earthdistance";',
    );
    expect(fixedContents).toMatchSnapshot();
  });
  it('should be able to create a new file to fix case where file does not exist yet', async () => {
    const declaredFileContents = await readFile(`${__dirname}/.extensions.sql`); // grab the declared contents
    const { contents: fixedContents } = await fix(null, {
      declaredFileContents,
    } as any);
    expect(fixedContents).toMatchSnapshot();
  });
});
