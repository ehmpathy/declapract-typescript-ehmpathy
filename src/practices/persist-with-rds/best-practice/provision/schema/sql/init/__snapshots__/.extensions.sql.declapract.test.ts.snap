// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.extensions.sql fix should be able to create a new file to fix case where file does not exist yet 1`] = `
"/**
 * postgres databases use extensions to increase functionality
 * - we have a minimum set of extensions that are required for sql-schema-generator
 * - services may also include their own extensions if they want additional functionality
 * - this is not managed by sql-schema-control
 *   - requires superuser privileges, cicd user only has db_owner privileges
 * - this is applied
 *   - automatically for integration-test db (through \`provision/docker/.../init/\` directory, with files cp'd by npm provision scripts)
 *   - by the deployment script for live deployments
 */

-- extensions required for all sql-schema-generator resources:
CREATE EXTENSION IF NOT EXISTS \\"uuid-ossp\\";
CREATE EXTENSION IF NOT EXISTS \\"pgcrypto\\";

-- extensions required for custom use cases in this service:
-- CREATE EXTENSION IF NOT EXISTS ..."
`;

exports[`.extensions.sql fix should be able to fix a file with custom extensions correctly 1`] = `
"/**
 * postgres databases use extensions to increase functionality
 * - we have a minimum set of extensions that are required for sql-schema-generator
 * - services may also include their own extensions if they want additional functionality
 * - this is not managed by sql-schema-control
 *   - requires superuser privileges, cicd user only has db_owner privileges
 * - this is applied
 *   - automatically for integration-test db (through \`provision/docker/.../init/\` directory, with files cp'd by npm provision scripts)
 *   - by the deployment script for live deployments
 */

-- extensions required for all sql-schema-generator resources:
CREATE EXTENSION IF NOT EXISTS \\"uuid-ossp\\";
CREATE EXTENSION IF NOT EXISTS \\"pgcrypto\\";

-- extensions required for custom use cases in this service:
CREATE EXTENSION IF NOT EXISTS \\"cube\\";
CREATE EXTENSION IF NOT EXISTS \\"earthdistance\\";"
`;
